Build hook spec (#955)
dart-lang/native: pkgs/native_assets_builder/doc/protocol/v1.md

Build Hook Specification
========================

Version 1: As of 2024-01-24 implemented behind `--enable-experiment=native-assets`

### Concepts

#### Asset

> A file that is identified by an `assetId`. There may be multiple files with the

Reviewer: ```suggestion
A reference to a file that is identified by an `assetId`. There may be multiple files with the
```

> A file that is identified by an `assetId`. There may be multiple files with the

Reviewer: I would try to separate the structure and the properties of the structure - so `an asset is a reference to a file, with additional information such as an ID, the build target, or the way it is linked.` and then `the ID is not necessarily unique` etc.

Author: I took a stab at this, but I'm not entirely sure what you want here. Feel free to push commits to this branch.

same `assetId` with different characteristics such as `target` or `link_mode`.

> Note that in v1 there are _only_ native code assets, no data assets.

Reviewer: What is a native code asset? What is a data asset?

#### AssetId
An asset must have an `assetId`. Dart code that uses an asset, references the
asset using the `assetId`.

A package MUST prefix all assetIds it defines with: `package:<package>/`, this
ensures assets don't conflict between packages.

Conventionally, an asset referenced from `lib/src/foo.dart` in `package:foo`
has `assetId = 'package:foo/src/foo.dart'`.

#### Target

> A target specifies operating system and architecture.

Reviewer: ```suggestion
A target specifies the operating system and architecture of the targeted platform.
```

List of `platform` and `architecture` tuples supported by the
Dart and Flutter SDKs to be specified below:
 * `linux_x64`
 * TODO: enumerate full list of supported platform/architecture tuples.

### `build.dart`

> A package with native assets MUST define a `build.dart` file which accepts

Reviewer: What is `A package with native assets` - if not a package which declares a `build.dart` file?

> a `--config` option as follows:

Reviewer: Maybe try to separate the contents from the syntax - so what does the build.dart accept as arguments, and what are the contents of those arguments?

```console
$ dart build.dart --config <build_config.yaml>
```

The `--config` option specifies location of the `build_config.yaml` file.

The `build.dart` file MUST:
 * Read the `build_config.yaml` file,
 * Build assets using configuration from `build_config.yaml`.
   * If `dry_run: true` in `build_config.yaml`, then this may be skipped.
 * Write assets into `out_dir` from `build_config.yaml`.
   * If `dry_run: true` in `build_config.yaml`, then this may be skipped.

>    * Filename are unrelated to `assetId`.

Reviewer: What does this mean?

Author: The file name inside out_dir/ doesn't really matter, because in Dart code you'll always access things by assetId. So in some sence the file name is irrelevant.

   * Arbitrary file names are fine, `build_output.yaml` will map `assetId` to files. 
   * MUST avoid file name `build_output.yaml`.
 * Write `build_output.yaml` into `out_dir` (from `build_config.yaml`).
   * This maps `assetId` to assets previous written in `out_dir`.
   * There may be multiple assets for a given `assetId` depending on
     characteristics like `target`, `link_mode`, etc.
   * If `dry_run: true` in `build_config.yaml`, the list of assets that would be
     generated must this be enumerated. Mapping `assetId`s to non-existing files
     is expected.

### `build_config.yaml`

> Configuration file passed to `build.dart` as the `--config` option.

Reviewer: Where does this file come from? Who writes this vs. the `build.dart` file?

dependency_metadata:
  # package name of direct dependency.
  some_package_name:
    # key value pairs.
    some_key: some_value

# Preferred link mode
link_mode_preference: dynamic | static | prefer-dynamic | prefer-static

# Path to output directory where assets should be placed.
#
# This is also where `build_output.yaml` should be written.
#
# Remark: Avoid using the name "build_output.yaml" for an asset file, this is
#         forbidden.
out_dir: /absolute/path/to/out_dir/

# Name of the package that contains the `build.dart`
#
# Remark: This is entirely redundant since this is a config file specified to
# `build.dart`, and the author of `build.dart` probably knows the name of the
# package they are writing.
package_name: my_package_with_native_assets

# Path to root folder for the package that contains `build.dart`.
#
# This is useful if `build.dart` wishes to find source code files embedded in
# its own package and compile them to an asset.
#
# Note that this will be most likely a path in the pub cache when the package
# with a `build.dart` is a dependency of another package.
package_root: /absolute/path/to/my_package_with_native_assets

# Target architecture
#

> # Combined with `target_os` this specifies the "target" for which assets

Reviewer: For v2: If this is a combination, then also declare it that way here?
```
target:
  architecture:
  os:
```

# should be built.
#
# Not provided on dry runs.
target_architecture: x64 | ia32 | arm | arm64 | riscv32 | riscv64

# Target operating system
#
# Combined with `target_architecture` this specifies the "target" for which
# assets should be built.
target_os: android | ios | linux | macos | windows

> # Version of this file.

Reviewer: Version of the syntax of the file or of the file?

version: 1.0.0
```

### `build_output.yaml`
Mapping from `assetId`s to files created by `build.dart`.

```yaml
# The list of assets.
#
# In dry runs, must contain assets for each architecture for the requested os.
assets:
  - id: 'package:my_package_with_native_assets/src/foo.dart'
    link_mode: dynamic | static | prefer-dynamic | prefer-static
    path:
      path_type: absolute | system | process | executable
      # Only provided for path_type absolute and system.
      #
      # If path_type absolute: The absolute path to the file name.
      #
      # If path_type system: The path of the dynamic library as available on 
      # the target machine's PATH.
      uri: /absolute/path/to/outdir/arbitrary_filename.whatever
    target: linux_x64
  ...

# The files used by this build.
#

> # If any of the files in [dependencies2] are modified after [timestamp], the

Reviewer: ```suggestion
# If any of the files in [dependencies] are modified after [timestamp], the
```

# build will be re-run.
#
# Not output on dry runs.
dependencies:
  - /absolute/path/to/my_package_with_native_assets/build.dart
  ...

# The time the build this output is for started.
#
# Must be before any of the files in dependencies are read.
timestamp: 2024-01-02 17:05:35.000

# Metadata usable for build.dart of packages directly depending on this package.
#
# Not output in dry runs.
metadata:
  # Key value pairs.
  some_key: some_value

# Version of this file.
version: 1.0.0
```

### `.dart_tool/native_assets.yaml`

This file is internal to the Dart SDK and Flutter SDK.
It produced by `native_asset_builder` (as embedded in the SDKs).
For each `target` it maps from `assetId` to `url`

> (with `urlKind` indicating if it's a relative or absolute path).

Reviewer: Again separate implementation from contents - what is `urlKind`?

Author: I'm not sure that we should have documentation about `.dart_tool/native_assets.yaml` in here. It's an internal vm format that happens to have some similarities with the native code assets.

See below, it's the path_type. And maybe that should have a different name. I'm open to suggestions. In the context of this file it is:

* absolute paths are absolute paths on the system where the Dart executable is running. This path type only really makes sense within the context of running things in JIT mode on a developers' machine, or in an Android app where the root path is the Android bundle.
* relative paths are relative to the kernel or aot snapshot. This makes sense if you're shipping a bundle through `dart build` and your asset is placed in a folder relative to the aotsnapshot.
* system paths are the same as for the protocol above.
* process "paths" have no path, and are the same as for the protocol above. 
* executable "paths" have no path, and are the same as for the protocol above.

```yaml
# Version of this file.
format-version: [1, 0, 0]

# A mapping from target and assetId to an asset.
native-assets:
  # A <target>.
  linux_x64:
    # An <assetId>.
    : 'package:my_package_with_native_assets/src/foo.dart'
      # A list of path_type and optionally a path.
      - absolute | relative | system | process | executable
      - <url> # Only provided for absolute, relative, and system path types.
  ...
```

This file is used by the embedded SDK to resolve `assetId`s when running in
development mode or build a deployable application (or standalone executable).

## Wishes to improve for v2

* Multiple asset types. Rename `Asset` to `NativeCodeAsset` and introduce `DataAsset`s.

> * The path_types really only make sense for `link_mode: dynamic`.

Reviewer: Maybe just drop `link_mode` for now completely? It is always dynamic, right?

Author: True, I haven't gotten to revive https://dart-review.googlesource.com/c/sdk/+/251263 yet.

Author: (Though this doc is describing the current protocol, so we can't drop it.)

* Fix the mismatch between `target` for output but `os` and `architecture` for config.
  * Should the architecture field of native code assets be optional and be ommitted in dry runs?

> * Introduce a `link.dart` protocol with a link input and link output.

Reviewer: I would rather have a v1 of that as well. To be discussed elsewhere though.

Author: Because what we do with `link.dart` influences `build.dart`, I think we need a holistic design doc. (Does the build.dart output an `AssetForLink`, or `List<Asset> assetsForLinking`, do we introduce unlimited intermediary asset types for linking, or that can subtype data assets. etc.)

An asset has different ways of being accessible in the final application. It is either brought in "manually" by having the package developer specify a path of the asset on the current system, it can be part of the Dart or Flutter SDK, or it can be already present in the target system. If the asset is bundled "manually", the Dart or Flutter SDK will take care of copying the asset from its specified location on the current system into the application bundle.

Assets are also called "native assets" to differentiate them from the Dart code also bundled with an application.

#### AssetId
An asset must have a string identifier called `assetId`. Dart code that uses an asset, references the asset using this `assetId`.

A package must prefix all `assetId`s it defines with: `package:<package>/`, `<package>` being the current package's name. This ensures assets don't conflict between packages.

Additionally, the convention is that an asset referenced from `lib/src/foo.dart` in `package:foo` has the `assetId` `'package:foo/src/foo.dart'`.

#### Target

A target specifies the operating system and architecture of the targeted platform of the application.

#### `build.dart`

> If there are assets in a package which have to be "manually" bundled, i.e. they are not shipped as part of Dart or Flutter and are also not present on the target system by default, that package must define a `build.dart` script specifying these assets. This script takes as input information such as the target OS and architecture in a `build_config.yaml` file, and returns as output a list of assets to be bundled in a `build_output.yaml` file. 

Author: Also the assets which are on the system or in Dart/Flutter must be declared in `build.dart` in the `build_output.yaml`. However, `build.dart` will not write a file for them to the output dir.

Build Hook Specification
========================

Version 1: As of 2024-01-24 implemented behind `--enable-experiment=native-assets`

### Concepts

#### Asset

An asset is a file together with metadata that tells the Dart and Flutter SDK how to bundle such asset in an application bundle.

Currently, all such assets are dynamic libraries.

> An asset must have an `assetId`. Dart code that uses an asset, references the asset using the `assetId`.

Author: Right, we have conflated the `Asset` class that _is_ metadata, and the file. This probs requires a full rewrite.

Since we also have a "metadata" field in the build input and build output, we shoudl probably refrain from calling everything that is _not_ the file metadata. Maybe we should call the file the "asset file".

> An asset must have a target os and target architecture. When accessing the asset at runtime, the asset for the current os and architecture is accessed.

Author: Not as input, but there is for the output. We should probably fix that in v2.

An asset must have a link mode. Currently, the only supported link mode is dynamic. (Static linking of native code will be added later.)

An asset must specify a path type for dynamic linking. 

* `absolute` paths are absolute paths in the output dir (see below).
* `system` paths are expected to resolve on the target machine PATH. In this case the asset is not a file but only metadata.

> * `process` "paths" have no path, symbols are resolved in the current process. In this case the asset is not a file but only metadata.

Author: Yeah so code assets are already much more involved than "a file". Code assets can be metadata only. And some variants of that metadata do not require any path.

path_type is probably not the right name.

Maybe it should be dynamic_library_type. But the executable and process one don't even have a dynamic library.

So then it should be dynamic_linking_type. This also sounds weird, because the linking always is the same in a way.

Maybe it should be dynamic_resolution_type. That gets closer. But the problem here is that we're also abstracting over the resolution partially. Dart and Flutter repackage dynamic libraries. So in this protocol for absolute paths, it's the absolute path at the point of this protocol. But for system paths, process, and executable the metadata gets passed in unmodified.

So then it should maybe be something like:

```yaml
- assetId: ...
  dynamic_resolution:
    type: absolute
  path: ... # Not nested in dynamic_resolution, it's the path for the protocol.
- assetId: ...
  dynamic_resolution:
    type: system
    path: ... # Nested in dynamic_resolution, it's the metadata to be passed into linking.
- assetId: ...
  dynamic_resolution:
    type: process
- assetId: ...
  dynamic_resolution:
    type: executable
```

wdyt?

* `executable` "paths" have no path, symbols are resolved in the current executable. In this case the asset is not a file but only metadata.

#### AssetId
An asset must have an `assetId`. Dart code that uses an asset, references the asset using the `assetId`.

A package must prefix all assetIds it defines with: `package:<package>/`, this ensures assets don't conflict between packages.

> Conventionally, an asset referenced from `lib/src/foo.dart` in `package:foo` has `assetId = 'package:foo/src/foo.dart'`.

Author: Hm, we don't have a convention yet.

Maybe:

* 'package:foo/src/foo.dart/some_name' if you only use it in that file.
* 'package:foo/some_name' if you use the asset in multiple dart files.

#### Target os and architecture

A target specifies the operating system and architecture of the targeted platform.

### `build.dart`
To bundle native assets with a package, that package must define a `build.dart`

> script which accepts a `--config` option as follows:

Reviewer: Technically, this is how to pass the config option, not how the build.dart accepts it. It accepts it by parsing the `List<String> args` and looking for a `--config` option.

```console
$ dart build.dart --config <build_config.yaml>
```

> The `--config` option specifies location of the `build_config.yaml` file.

Reviewer: ```suggestion
The `--config` option specifies the location of the `build_config.yaml` file.
```
How? As an absolute path?

The `build.dart` file MUST:
 * Read the `build_config.yaml` file,

>  * Build assets using configuration from `build_config.yaml`.

Reviewer: ```suggestion
 * Build assets using the configuration from `build_config.yaml`.
```

>    * If `dry_run: true` in `build_config.yaml`, then this may be skipped.

Reviewer: So its not a MUST after all?

Author: Correct.

With a linear doc it's a bit tricky to introduce concepts and then later nuance them.
(It feels slightly easier to do with API docs where you can make lots of circular references. But of course there also the docs on the main class should also be kind of the first entry point.)

Open to suggestions. (Please feel free to push commits to rewrite this instead of leaving comments.)

>  * Write assets into `out_dir` from `build_config.yaml`.

Reviewer: What is `out_dir`?

>    * If `dry_run: true` in `build_config.yaml`, then this may be skipped.

Reviewer: ```suggestion
   * If the boolean parameter `dry_run` is set to true in the build configuration, then this may be skipped.
```

>    * Filename are unrelated to `assetId`.

Reviewer: ```suggestion
   * The file name inside `out_dir` doesn't really matter, because in Dart code you'll always access things by `assetId`. So in some sense the file name is irrelevant.
```

>    * Arbitrary file names are fine, `build_output.yaml` will map `assetId` to files. 

Reviewer: What does this mean?

Author: The file names of the assets don't really matter. It could be file1.foo, file2.foo, file3.foo. Because the only way assets are accessed from Dart code is through their assetId.

However, we do want to have to files written to disk rather than having a buffer of bytes for files, because quite often the files are produces by an external `Process` call or for data assets already in your package.

>    * MUST avoid file name `build_output.yaml`.

Reviewer: As a name for an asset?

Author: If you write your asset to out_dir/build_output.yaml it's going to fail to write build_output.yaml

Maybe in v2 we should consider writing the assets somewhere else than the build output.

 * Write `build_output.yaml` into `out_dir` (from `build_config.yaml`).

>    * This maps `assetId` to assets previous written in `out_dir`.

Reviewer: ```suggestion
   * This maps the `assetId`s to assets previously written into `out_dir`.
```
I am not sure what `maps` means.

>    * There may be multiple assets for a given `assetId` depending on

Reviewer: You might want to add something about the uniqueness of an `assetId` - this was unclear in an earlier review comment as well.

     characteristics like `target`, `link_mode`, etc.

>    * If `dry_run: true` in `build_config.yaml`, the list of assets that would be

Reviewer: How do you enumerate a list? I don't understand what that means.
